File: src/lib/db/repositories/variantRepository.ts
import { and, eq } from "drizzle-orm";

import type { AppDatabase } from "@/lib/db/client";
import { buildVariants, variantSetAttachments } from "@/lib/db/schema";
import {
  parseSubclassKitJson,
  serializeSubclassKit,
  type SubclassKitFields,
} from "@/lib/builds/subclassKit";

export type VariantRecord = {
  id: string;
  buildId: string;
  name: string;
  isDefault: boolean;
  exoticWeaponHash: number | null;
  exoticWeaponName: string | null;
  artifactHash: number | null;
  artifactName: string | null;
  artifactConfig: number[];
  /** Null = legacy: fall back to kit fields on build.subclass. */
  subclassKit: SubclassKitFields | null;
  notes: string | null;
  createdAt: string;
  updatedAt: string;
};

function parseArtifactConfig(raw: string | null | undefined): number[] {
  if (!raw) return [];
  try {
    const parsed = JSON.parse(raw) as unknown;
    return Array.isArray(parsed) ? parsed.filter((n): n is number => typeof n === "number") : [];
  } catch {
    return [];
  }
}

export type AttachmentRecord = {
  id: string;
  variantId: string;
  setId: string;
  mode: "live" | "snapshot";
  snapshotConfigs: SnapshotConfig[] | null;
  attachedAt: string;
};

export type SnapshotConfig = {
  slot: string;
  itemHash: number;
  itemName: string;
  selectedPerks?: number[];
  masterworkHash?: number | null;
  modHashes?: number[] | null;
  instanceId?: string | null;
};

function rowToVariant(row: typeof buildVariants.$inferSelect): VariantRecord {
  return {
    id: row.id,
    buildId: row.buildId,
    name: row.name,
    isDefault: row.isDefault === 1,
    exoticWeaponHash: row.exoticWeaponHash,
    exoticWeaponName: row.exoticWeaponName,
    artifactHash: row.artifactHash ?? null,
    artifactName: row.artifactName ?? null,
    artifactConfig: parseArtifactConfig(row.artifactConfig),
    subclassKit: parseSubclassKitJson(row.subclassKit),
    notes: row.notes,
    createdAt: row.createdAt,
    updatedAt: row.updatedAt,
  };
}

function rowToAttachment(row: typeof variantSetAttachments.$inferSelect): AttachmentRecord {
  return {
    id: row.id,
    variantId: row.variantId,
    setId: row.setId,
    mode: row.mode as "live" | "snapshot",
    snapshotConfigs: row.snapshotConfigs ? (JSON.parse(row.snapshotConfigs) as SnapshotConfig[]) : null,
    attachedAt: row.attachedAt,
  };
}

export function listVariants(db: AppDatabase, buildId: string): VariantRecord[] {
  return db
    .select()
    .from(buildVariants)
    .where(eq(buildVariants.buildId, buildId))
    .all()
    .map(rowToVariant);
}

export function getVariant(db: AppDatabase, buildId: string, variantId: string): VariantRecord | null {
  const row = db
    .select()
    .from(buildVariants)
    .where(and(eq(buildVariants.id, variantId), eq(buildVariants.buildId, buildId)))
    .get();
  return row ? rowToVariant(row) : null;
}

export function createVariantRecord(
  db: AppDatabase,
  input: {
    id: string;
    buildId: string;
    name: string;
    isDefault?: boolean;
    exoticWeaponHash?: number | null;
    exoticWeaponName?: string | null;
    artifactHash?: number | null;
    artifactName?: string | null;
    artifactConfig?: number[];
    subclassKit?: SubclassKitFields | null;
    notes?: string | null;
    now: string;
  },
): VariantRecord {
  db.insert(buildVariants)
    .values({
      id: input.id,
      buildId: input.buildId,
      name: input.name,
      isDefault: input.isDefault ? 1 : 0,
      exoticWeaponHash: input.exoticWeaponHash ?? null,
      exoticWeaponName: input.exoticWeaponName ?? null,
      artifactHash: input.artifactHash ?? null,
      artifactName: input.artifactName ?? null,
      artifactConfig: JSON.stringify(input.artifactConfig ?? []),
      subclassKit:
        input.subclassKit != null ? serializeSubclassKit(input.subclassKit) : null,
      notes: input.notes ?? null,
      createdAt: input.now,
      updatedAt: input.now,
    })
    .run();
  return getVariant(db, input.buildId, input.id)!;
}

export function updateVariantRecord(
  db: AppDatabase,
  buildId: string,
  variantId: string,
  patch: {
    name?: string;
    exoticWeaponHash?: number | null;
    exoticWeaponName?: string | null;
    artifactHash?: number | null;
    artifactName?: string | null;
    artifactConfig?: number[];
    subclassKit?: SubclassKitFields | null;
    notes?: string | null;
    now: string;
  },
): VariantRecord | null {
  const existing = getVariant(db, buildId, variantId);
  if (!existing) return null;

  db.update(buildVariants)
    .set({
      name: patch.name ?? existing.name,
      exoticWeaponHash: patch.exoticWeaponHash !== undefined ? patch.exoticWeaponHash : existing.exoticWeaponHash,
      exoticWeaponName:
        patch.exoticWeaponName !== undefined ? patch.exoticWeaponName : existing.exoticWeaponName,
      artifactHash: patch.artifactHash !== undefined ? patch.artifactHash : existing.artifactHash,
      artifactName: patch.artifactName !== undefined ? patch.artifactName : existing.artifactName,
      artifactConfig:
        patch.artifactConfig !== undefined
          ? JSON.stringify(patch.artifactConfig)
          : JSON.stringify(existing.artifactConfig),
      subclassKit:
        patch.subclassKit !== undefined
          ? patch.subclassKit == null
            ? null
            : serializeSubclassKit(patch.subclassKit)
          : existing.subclassKit != null
            ? serializeSubclassKit(existing.subclassKit)
            : null,
      notes: patch.notes !== undefined ? patch.notes : existing.notes,
      updatedAt: patch.now,
    })
    .where(and(eq(buildVariants.id, variantId), eq(buildVariants.buildId, buildId)))
    .run();

  return getVariant(db, buildId, variantId);
}

/** Wipe kit on every variant of a build (tree change baseline). */
export function wipeAllVariantKits(
  db: AppDatabase,
  buildId: string,
  kit: SubclassKitFields,
  now: string,
): void {
  const variants = listVariants(db, buildId);
  for (const v of variants) {
    updateVariantRecord(db, buildId, v.id, { subclassKit: kit, now });
  }
}

export function deleteVariantRecord(db: AppDatabase, buildId: string, variantId: string): boolean {
  const result = db
    .delete(buildVariants)
    .where(and(eq(buildVariants.id, variantId), eq(buildVariants.buildId, buildId)))
    .run();
  return result.changes > 0;
}

export function listAttachments(db: AppDatabase, variantId: string): AttachmentRecord[] {
  return db
    .select()
    .from(variantSetAttachments)
    .where(eq(variantSetAttachments.variantId, variantId))
    .all()
    .map(rowToAttachment);
}

export function replaceAttachments(
  db: AppDatabase,
  variantId: string,
  attachments: Array<Omit<AttachmentRecord, "id" | "variantId" | "attachedAt"> & { id?: string }>,
  now: string,
): AttachmentRecord[] {
  db.delete(variantSetAttachments).where(eq(variantSetAttachments.variantId, variantId)).run();

  for (const attachment of attachments) {
    db.insert(variantSetAttachments)
      .values({
        id: attachment.id ?? crypto.randomUUID(),
        variantId,
        setId: attachment.setId,
        mode: attachment.mode,
        snapshotConfigs: attachment.snapshotConfigs ? JSON.stringify(attachment.snapshotConfigs) : null,
        attachedAt: now,
      })
      .run();
  }

  return listAttachments(db, variantId);
}

File: flutter/packages/db/lib/src/repos/variant_repository.dart
import 'package:drift/drift.dart';

import '../app_database.dart';
import '../tables.dart';
import 'json_codec.dart';
import 'library_records.dart';

VariantRecord _rowToVariant(BuildVariant row) {
  return VariantRecord(
    id: row.id,
    buildId: row.buildId,
    name: row.name,
    isDefault: row.isDefault == 1,
    exoticWeaponHash: row.exoticWeaponHash,
    exoticWeaponName: row.exoticWeaponName,
    artifactHash: row.artifactHash,
    artifactName: row.artifactName,
    artifactConfig: parseIntJsonArray(row.artifactConfig),
    subclassKit: decodeJsonValue(row.subclassKit),
    notes: row.notes,
    createdAt: row.createdAt,
    updatedAt: row.updatedAt,
  );
}

AttachmentRecord _rowToAttachment(VariantSetAttachment row) {
  return AttachmentRecord(
    id: row.id,
    variantId: row.variantId,
    setId: row.setId,
    mode: row.mode,
    snapshotConfigs: parseSnapshotConfigsJson(row.snapshotConfigs),
    attachedAt: row.attachedAt,
  );
}

Future<List<VariantRecord>> listVariants(
  AppDatabase db,
  String buildId,
) async {
  final rows = await (db.select(db.buildVariants)
        ..where((t) => t.buildId.equals(buildId)))
      .get();
  return rows.map(_rowToVariant).toList();
}

Future<VariantRecord?> getVariant(
  AppDatabase db,
  String buildId,
  String variantId,
) async {
  final row = await (db.select(db.buildVariants)
        ..where(
          (t) => t.id.equals(variantId) & t.buildId.equals(buildId),
        ))
      .getSingleOrNull();
  return row == null ? null : _rowToVariant(row);
}

Future<VariantRecord> createVariantRecord(
  AppDatabase db, {
  required String id,
  required String buildId,
  required String name,
  bool isDefault = false,
  int? exoticWeaponHash,
  String? exoticWeaponName,
  int? artifactHash,
  String? artifactName,
  List<int> artifactConfig = const [],
  Object? subclassKit,
  String? notes,
  required String now,
}) async {
  await db.into(db.buildVariants).insert(
        BuildVariantsCompanion.insert(
          id: id,
          buildId: buildId,
          name: name,
          isDefault: Value(isDefault ? 1 : 0),
          exoticWeaponHash: Value(exoticWeaponHash),
          exoticWeaponName: Value(exoticWeaponName),
          artifactHash: Value(artifactHash),
          artifactName: Value(artifactName),
          artifactConfig: Value(encodeIntJsonArray(artifactConfig)),
          subclassKit: Value(encodeJsonValue(subclassKit ?? const <String, Object?>{})),
          notes: Value(notes),
          createdAt: now,
          updatedAt: now,
        ),
      );
  return (await getVariant(db, buildId, id))!;
}

Future<VariantRecord?> updateVariantRecord(
  AppDatabase db,
  String buildId,
  String variantId, {
  String? name,
  Value<int?> exoticWeaponHash = const Value.absent(),
  Value<String?> exoticWeaponName = const Value.absent(),
  Value<int?> artifactHash = const Value.absent(),
  Value<String?> artifactName = const Value.absent(),
  List<int>? artifactConfig,
  Value<Object?> subclassKit = const Value.absent(),
  Value<String?> notes = const Value.absent(),
  required String now,
}) async {
  final existing = await getVariant(db, buildId, variantId);
  if (existing == null) return null;

  await (db.update(db.buildVariants)
        ..where(
          (t) => t.id.equals(variantId) & t.buildId.equals(buildId),
        ))
      .write(
    BuildVariantsCompanion(
      name: Value(name ?? existing.name),
      exoticWeaponHash: exoticWeaponHash.present
          ? exoticWeaponHash
          : Value(existing.exoticWeaponHash),
      exoticWeaponName: exoticWeaponName.present
          ? exoticWeaponName
          : Value(existing.exoticWeaponName),
      artifactHash:
          artifactHash.present ? artifactHash : Value(existing.artifactHash),
      artifactName:
          artifactName.present ? artifactName : Value(existing.artifactName),
      artifactConfig: Value(
        encodeIntJsonArray(artifactConfig ?? existing.artifactConfig),
      ),
      subclassKit: subclassKit.present
          ? Value(encodeJsonValue(subclassKit.value ?? const <String, Object?>{}))
          : Value(encodeJsonValue(existing.subclassKit ?? const <String, Object?>{})),
      notes: notes.present ? notes : Value(existing.notes),
      updatedAt: Value(now),
    ),
  );
  return getVariant(db, buildId, variantId);
}

Future<bool> deleteVariantRecord(
  AppDatabase db,
  String buildId,
  String variantId,
) async {
  final n = await (db.delete(db.buildVariants)
        ..where(
          (t) => t.id.equals(variantId) & t.buildId.equals(buildId),
        ))
      .go();
  return n > 0;
}

Future<List<AttachmentRecord>> listAttachments(
  AppDatabase db,
  String variantId,
) async {
  final rows = await (db.select(db.variantSetAttachments)
        ..where((t) => t.variantId.equals(variantId)))
      .get();
  return rows.map(_rowToAttachment).toList();
}

/// Replace all attachments for [variantId] with [attachments].
///
/// When [AttachmentWrite.id] is omitted, allocates `$variantId-att-N` that
/// does not collide with ids already claimed by sibling writes (BUG-20260726-016:
/// re-attach replace-by-type kept `…-att-0` while the new row also got `…-att-0`).
Future<List<AttachmentRecord>> replaceAttachments(
  AppDatabase db,
  String variantId,
  List<AttachmentWrite> attachments,
  String now,
) async {
  await (db.delete(db.variantSetAttachments)
        ..where((t) => t.variantId.equals(variantId)))
      .go();

  final usedIds = <String>{
    for (final a in attachments)
      if (a.id != null && a.id!.isNotEmpty) a.id!,
  };
  var nextIndex = 0;
  String allocateId() {
    String candidate;
    do {
      candidate = '$variantId-att-${nextIndex++}';
    } while (usedIds.contains(candidate));
    usedIds.add(candidate);
    return candidate;
  }

  for (final a in attachments) {
    final id = (a.id != null && a.id!.isNotEmpty) ? a.id! : allocateId();
    await db.into(db.variantSetAttachments).insert(
          VariantSetAttachmentsCompanion.insert(
            id: id,
            variantId: variantId,
            setId: a.setId,
            mode: a.mode,
            snapshotConfigs: Value(encodeSnapshotConfigsJson(a.snapshotConfigs)),
            attachedAt: now,
          ),
        );
  }
  return listAttachments(db, variantId);
}

/// Input for [replaceAttachments].
class AttachmentWrite {
  const AttachmentWrite({
    this.id,
    required this.setId,
    required this.mode,
    this.snapshotConfigs,
  });

  final String? id;
  final String setId;
  final String mode;
  final List<Map<String, Object?>>? snapshotConfigs;
}
